src/gtc/passes/oir_optimizations/inlining.py

```diff
@@ -0,0 +1,112 @@
+# -*- coding: utf-8 -*-
+#
+# GTC Toolchain - GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+import copy as cp
+from typing import Any, Dict, Optional
+
+from eve import NodeTranslator
+from gtc import oir
+
+
+class MaskInlining(NodeTranslator):
```

Comment:
I think the pass would be much more expressive if it would be split into 2 visitors, first detect inline candidates and a second, probably very simple one, to actually do the replacements. I cannot easily see here which preconditions for applying the merge are checked. Please add a docstring here that explains under which conditions you do merging. And the concrete question: do you check if fields that are part of the mask are not modified in the body? That's the reason we have the separate mask field in the first place.

Reply:
I think https://github.com/GridTools/gt4py/blob/73ab040f9201cfe36cccd358efdc226a5f569c75/src/gtc/passes/oir_optimizations/horizontal_execution_merging.py is a good source for inspiration.